program/rust/src/time_machine_types.rs

```diff
@@ -6,20 +6,375 @@ use crate::c_oracle_header::{
     PC_PRICE_T_COMP_OFFSET,
 };
 use crate::error::OracleError;
+use crate::utils::try_convert;
 use bytemuck::{
     Pod,
     Zeroable,
 };
-use solana_program::msg;
+use std::cmp;
 
+//To make it easy to change the types to allow for more usefull
+//running sums if needed
+use {
+    i64 as SignedTrackerRunningSum,
+    u64 as UnsignedTrackerRunningSum,
+};
+//multiply by this to make up for errors that result when dividing integers
+pub const SMA_TRACKER_PRECISION_MULTIPLIER: i64 = 10;
+
+
+pub trait Tracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    ///initializes a zero initialized tracker
+    fn initialize(&mut self) -> Result<(), OracleError>;
+
+    ///add a new price to the tracker
+    fn add_price(
+        &mut self,
+        prev_time: i64,
+        prev_price: i64,
+        prev_conf: u64,
+        current_time: i64,
+        current_price: i64,
+        current_conf: u64,
+    ) -> Result<(), OracleError>;
+    ///gets the index of the next entry
+    fn get_next_entry(current_entry: usize) -> usize {
+        (current_entry + 1) % NUM_ENTRIES
+    }
+    ///Given a time, return the entry of that time
+    fn time_to_entry(time: i64) -> Result<usize, OracleError> {
+        Ok(try_convert::<i64, usize>(time / GRANUALITY)? % NUM_ENTRIES)
+    }
+    ///returns the remining time before the next multiple of GRANUALITY
+    fn get_time_to_entry_end(time: i64) -> i64 {
+        (GRANUALITY - (time % GRANUALITY)) % GRANUALITY
+    }
+}
+
+#[derive(Debug, Copy, Clone)]
+#[repr(C)]
+/// Represents an SMA Tracker that has NUM_ENTRIES entries
+/// each tracking time weighted sums for GRANUALITY seconds periods.
+///The prices are assumed to be provided under some fixed point representation, and the computation
+/// gurantees accuracy up to the last decimal digit in the fixed point representation.
+/// Assumes THRESHOLD < GRANUALITY
+pub struct SmaTracker<const GRANUALITY: i64, const NUM_ENTRIES: usize, const THRESHOLD: i64> {
+    valid_entry_counter:        u64, // is incremented everytime we add a valid entry
+    max_update_time_gap:        i64, // maximum time between two updates in the current entry.
```

Comment:
Could you add an explanation of the semantics of this field? At a first glance, it could be interpreted as either a limit or a measurement, and it's only when you read the implementation it becomes clear it's a measurement.